libraries/hotstuff/include/eosio/hotstuff/qc_chain.hpp

```diff
@@ -77,71 +77,261 @@ namespace eosio::hotstuff {
+      explicit pending_quorum_certificate(const fc::sha256& proposal_id,
+                                          const digest_type& proposal_digest,
+                                          size_t num_finalizers,
+                                          size_t quorum) :
+         pending_quorum_certificate(num_finalizers, quorum) {         
+         _proposal_id = proposal_id;
+         _proposal_digest.assign(proposal_digest.data(), proposal_digest.data() + 32);
+         _state = state_t::unrestricted;
       }
 
+      size_t num_weak()   const { return _weak_votes.count(); }
+      size_t num_strong() const { return _strong_votes.count(); }
+
+      bool   valid() const { return _state >= state_t::weak_achieved; }
+
+      // ================== begin compatibility functions =======================
+      // these assume *only* strong votes
+      
+      // this function is present just to make the tests still work
+      // it will be removed, as well as the _proposal_id member of this class
       quorum_certificate_message to_msg() const {
-         return {.proposal_id = proposal_id,
-                 .active_finalizers = [this]() {
-                           std::vector<unsigned_int> r;
-                           r.resize(active_finalizers.num_blocks());
-                           boost::to_block_range(active_finalizers, r.begin());
-                           return r;
-                        }(),
-                 .active_agg_sig = active_agg_sig};
+         return {.proposal_id    = _proposal_id,
+                 .strong_votes   = bitset_to_vector(_strong_votes._bitset),
+                 .active_agg_sig = _strong_votes._sig};
+      }
+
+      bool                 is_quorum_met() const { return valid(); }
+      const fc::sha256&    get_proposal_id() const { return _proposal_id; }
+      std::string          get_votes_string() const {
+         return std::string("strong(\"") + bitset_to_string(_strong_votes._bitset) + "\", weak(\"" +
+            bitset_to_string(_weak_votes._bitset) + "\"";
+      }
+      // ================== end compatibility functions =======================
+
+      void reset(const fc::sha256& proposal_id, const digest_type& proposal_digest, size_t num_finalizers, size_t quorum) {
+         _proposal_id = proposal_id;
+         _proposal_digest.assign(proposal_digest.data(), proposal_digest.data() + 32);
+         _quorum = quorum;
+         _strong_votes.reset(num_finalizers);
+         _weak_votes.reset(num_finalizers);
+         _num_finalizers = num_finalizers;
+         _state = state_t::unrestricted;
       }
 
-      void reset(const fc::sha256& proposal, size_t finalizer_size) {
-         proposal_id = proposal;
-         active_finalizers = hs_bitset{finalizer_size};
-         active_agg_sig = fc::crypto::blslib::bls_signature();
-         quorum_met = false;
+      bool add_strong_vote(const std::vector<uint8_t>& proposal_digest,
```

Comment:
Why this and other implementation not in .cpp file?

Reply:
Why would it be much better in the .cpp file? 
In any case like this it is easier to modify for now, and if we want later we can move the implementations in the .cpp.